Nifi 2398 - Apache Ignite Get Processor

mans2singh opened the pull request:
Ignite get processor.

pvillard31 commented on nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/GetIgniteCache.java:
```diff
@@ -0,0 +1,119 @@
+
+    /** Flow file attribute keys and messages */
+    public static final String IGNITE_GET_FAILED_REASON_ATTRIBUTE_KEY = "ignite.cache.get.failed.reason";
+    public static final String IGNITE_GET_FAILED_MISSING_KEY_MESSAGE = "The FlowFile key attribute was missing";
```
Shouldn't it be a key instead of a string message here?

pvillard31 commented on nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/cache/GetIgniteCache.java:
```diff
@@ -0,0 +1,119 @@
+    public static final String IGNITE_GET_FAILED_MESSAGE_PREFIX = "The cache request failed because of ";
+
+    static {
+        descriptors = new ArrayList<>();
```
You are sharing a static reference for descriptors in both processors, it results in a race condition and not all properties are present. Have a look to other processors extending an abstract class to have an example.

mans2singh commented:
Hi @pvillard31 - Can you please point me to which attribute you are referring to since the constant (IGNITE_GET_FAILED_MISSING_KEY_MESSAGE) is a message not a key.
I will correct the 2nd issue that you mentioned.
Thanks

pvillard31 commented:
@mans2singh you are using the constant defined L62 in the annotation L55 as the attribute name. I believe this is a mistake but correct me if I'm wrong.
````java
@WritesAttribute(attribute = GetIgniteCache.IGNITE_GET_FAILED_MISSING_KEY_MESSAGE, ...
````

mans2singh commented:
@pvillard31 - I will work on the issues.  Thanks for your review.

mans2singh commented:
@pvillard31 - 

I've updated the code based on your comments.  Let me know if you have any other thoughts/recommendations.

Thanks

pvillard31 commented on nifi-nar-bundles/nifi-ignite-bundle/nifi-ignite-processors/src/main/java/org/apache/nifi/processors/ignite/AbstractIgniteProcessor.java:
```diff
@@ -92,6 +94,14 @@ public void initializeIgnite(ProcessContext context) {
+        List<Ignite> grids = Ignition.allGrids();
+
+        if ( grids.size() == 1 ) {
+            getLogger().warn("Ignite grid already avaialble");
```
I think this should not be a warning because if using Put and Get processors on the canvas we will get a warning bulletin although there is no issue. Maybe log as info and inform the user we use a shared grid? (also take care of the typo).

pvillard31 commented:
I have played with both Put and Get processors and it seems to work fine. However, when stopping/starting processors to change properties (to test batching and performances), I got in a situation with this exception:

````
2016-10-03 23:32:42,633 WARN [Timer-Driven Process Thread-4] o.a.n.c.t.ContinuallyRunProcessorTask Administratively Yielding PutIgniteCache[id=4d80623c-0157-1000-0000-0000179dcc8f] due to uncaught Exception: java.lang.IllegalStateException: Data streamer has been closed.
2016-10-03 23:32:42,634 WARN [Timer-Driven Process Thread-4] o.a.n.c.t.ContinuallyRunProcessorTask
java.lang.IllegalStateException: Data streamer has been closed.
	at org.apache.ignite.internal.processors.datastreamer.DataStreamerImpl.enterBusy(DataStreamerImpl.java:355) ~[na:na]
	at org.apache.ignite.internal.processors.datastreamer.DataStreamerImpl.addData(DataStreamerImpl.java:489) ~[na:na]
	at org.apache.nifi.processors.ignite.cache.PutIgniteCache.onTrigger(PutIgniteCache.java:274) ~[na:na]
	at org.apache.nifi.processor.AbstractProcessor.onTrigger(AbstractProcessor.java:27) ~[nifi-api-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.apache.nifi.controller.StandardProcessorNode.onTrigger(StandardProcessorNode.java:1060) ~[nifi-framework-core-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.apache.nifi.controller.tasks.ContinuallyRunProcessorTask.call(ContinuallyRunProcessorTask.java:136) [nifi-framework-core-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.apache.nifi.controller.tasks.ContinuallyRunProcessorTask.call(ContinuallyRunProcessorTask.java:47) [nifi-framework-core-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.apache.nifi.controller.scheduling.TimerDrivenSchedulingAgent$1.run(TimerDrivenSchedulingAgent.java:123) [nifi-framework-core-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_77]
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) [na:1.8.0_77]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) [na:1.8.0_77]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) [na:1.8.0_77]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_77]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_77]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_77]
````

It was not possible to get the flow back to OK unless restarting NiFi. I believe something is not closed correctly.

To reproduce, take this template, run all processors, everything should be OK, stop GetIgnite processor and restart it, do the same with PutIgnite. You should get the error.
Here is the template I use: https://gist.github.com/pvillard31/8a0c1d0344bf5bdfdf4cd460e936905c

mans2singh commented:
@pvillard31 - 

I've updated the code and checked it with your template.  I've also added more integration tests to check processors can be restarted multiple times.  Please let me know if you have any additional comments.

Thanks again

pvillard31 commented:
LGTM +1
Full build with contrib-check.
Ran multiple workflows with various configurations with a local Ignite instance. Everything worked as expected, problems raised before have been solved. Thanks for your work @mans2singh !

mans2singh commented:
Thanks @pvillard31 for your review and advice.  Mans
